Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,621 @@
+
+#### Alternation
+
+Alternations are used to match one of multiple possible patterns.
+
+```txt
+pattern ::= pattern '|' pattern
+```
+
+If there are one or more capturing groups within an alternation, the resulting
+capture type is an `Alternation` that's generic over each option's underlying
+pattern.
+
+```swift
+/([01]+)|[0-9]+|([0-9A-F]+)/
+// => `Regex<Alternation<(Substring, Void, Substring)>>`
+```
+
+If there are no capturing groups within an alternation the resulting capture
+type is `Void`.
+
+```swift
+/[01]+|[0-9]+|[0-9A-F]+/
+// => `Regex<Void>`
+```
+
+Nested captures follow the algebra previously described.
+
+```swift
+/([01]+|[0-9]+|[0-9A-F]+)/
+// => `Regex<Substring>`
+/(([01]+)|([0-9]+)|([0-9A-F]+))/
+// => `Regex<(Substring, Alternation<(Substring, Substring, Substring))>>`
+/(?<overall>(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+))/
+// => `Regex<(overall: Substring, Alternation<(binary: Substring, decimal: Substring, hex: Substring))>>`
+```
+
+At the use site, `Alternation` behaves like an `enum`. Ideally you should be
+able to exhaustively switch over all the captures.
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .map {
+        switch $0 {
+        case let .0(binary):
+            return Int(binary, radix: 2)
+        case let .1(decimal):
+            return Int(decimal, radix: 10)
+        case let .2(hex):
+            return Int(decimal, radix: 16)
+        }
+    }
+
+// Or with named captures:
+let number = line
+    .firstMatch(/(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0 {
+        case let .binary(str):
+            return Int(str, radix: 2)
+        case let .decimal(str):
+            return Int(str, radix: 10)
+        case let .hex(str):
+            return Int(str, radix: 16)
+        }
+    }
+```
+
+In the fullness of time, we'd like to design a language feature to support this.
+In the meantime, we would like to do the best we can and leave the door open for
+a source-compatible migration.
+
+With variadic generics we think we can define the following `Alternation` type.
+
+```swift
+@dynamicMemberLookup
+public struct Alternation<Captures> { ... }
+
+extension<Option...> Alternation where Captures == (Option...) {
+    public var options: (Option?...) { get }
+    
+    public subscript<T>(dynamicMember keyPath: KeyPath<(T?...), T>) -> T {
+      options[keyPath: keyPath]
+    }
+}
+```
+
+An optional projection property, `options`, presents all options as a tuple of
+optionals. The Standard Library will provide the runtime guarantee that only one
+element of the `options` tuple will be non-nil. The programmer can use a
+`switch` statement to pattern-match the tuple and handle each case, or directly
+access the tuple's properties via key path dynamic member lookup.
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .map {
```

Comment:
These were `flatMap` because `Int(_:radix:)` returns an `Int?`, so using just `map` on an `Optional` results in `Int??`

Reply:
Ah, I had assume `init(_:radix:)` returned an non-optional. Didn't look it up.